Matter density as a trait

preshanth opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address the creation of a matter density trait. The goal here is to allow pedagogical models.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes a part of #17150 

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

nstarman commented on astropy/cosmology/_src/flrw/base.py:
```diff
@@ -94,7 +95,14 @@ def quad(*args, **kwargs):
 
 
 @dataclass_decorator
-class FLRW(Cosmology, ScaleFactor, TemperatureCMB, _CriticalDensity, _BaryonComponent):
```
the `def Om()` method should be removed from here, so that it uses the one on `_MatterDensity`.

preshanth replied:
That was an oversight. Thanks for catching that. Its done.

nstarman reviewed (changes requested):
Thanks @preshanth for the PR! Some comments.

preshanth commented:
I ask because I can compute dark matter density as matter - baryons and I can do that inside the MatterDensity class rather than have a separate one.  I have nominally made a separate branch and moved the dark matter there too. I have moved the functions. If we agree on what formats. I am happy to do photon and dark energy along with distance measures next. I will leave the neutrinos to you.

nstarman commented:
> I was wondering what you thought about having all the matter densities in one class. Include the dark matter here too.

I think it's best to keep them separate. We want to use these traits to build e.g. `MatterOnlyCosmology`, for which we should have only $\Omega_{m,0}$, $\Omega_{m}(z)$.

nstarman reviewed (approved):
Thanks @preshanth. This looks great.
